roomserver/storage/sqlite3/state_block_table.go

```diff
@@ -32,228 +33,104 @@ import (
 
 const stateDataSchema = `
   CREATE TABLE IF NOT EXISTS roomserver_state_block (
-    state_block_nid INTEGER NOT NULL,
-    event_type_nid INTEGER NOT NULL,
-    event_state_key_nid INTEGER NOT NULL,
-    event_nid INTEGER NOT NULL,
-    UNIQUE (state_block_nid, event_type_nid, event_state_key_nid)
+    state_block_nid INTEGER PRIMARY KEY AUTOINCREMENT,
+	state_block_hash BLOB UNIQUE,
+    event_nids TEXT NOT NULL DEFAULT '[]'
   );
 `
 
-const insertStateDataSQL = "" +
-	"INSERT INTO roomserver_state_block (state_block_nid, event_type_nid, event_state_key_nid, event_nid)" +
-	" VALUES ($1, $2, $3, $4)"
-
-const selectNextStateBlockNIDSQL = `
-SELECT IFNULL(MAX(state_block_nid), 0) + 1 FROM roomserver_state_block
+const insertStateDataSQL = `
+	INSERT INTO roomserver_state_block (state_block_hash, event_nids)
+		VALUES ($1, $2)
+		ON CONFLICT (state_block_hash) DO UPDATE SET event_nids=$2
```

Comment:
If we get conflicts on the hash, it's because we've tried to insert a state block that is a duplicate of one we already have. In that case, we need to do `DO UPDATE` so that the `RETURNING` statement sends back the state block ID of the block that *already exists* so that we can reuse the existing NID, whereas `DO NOTHING` wouldn't return anything at all.